### PR TITLE
docs(config) document configuration cache for webpack v5

### LINKED ISSUE #2605: Document webpack change: add persistent caching prototype
*A pull request by @sokra was merged and maintainers requested a documentation change.*

See pull request: https://github.com/webpack/webpack/pull/8162

---

**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
yes

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
The `cache` options can now take an object of this form: 

``` js
{
  type: "filesystem",

  // directory where to store the cache
  // cache will be stored in subfolded `${name}-${mode}`
  cacheDirectory: "/path/to/dir",

  // subfolder name
  name: "something",

  // configuration version
  // cache is invalidated when this is updated
  version: "1xz2",

  // when to store the cache-able items
  // idle: while the compiler is idling
  // instant: while compiling, blocking
  // background: while compiling, non-blocking
  store: "idle", 

  // show warning or info messages
  loglevel: "warning",

  // use a custom hash algorithm i. e. for performance
  hashAlgorithm: "md4"
}
```

For descriptions see schema.

This will lead to (de)serialization of modules and generated assets to the filesystem.

### PR DESCRIPTION
**Pull request by EugeneHlushko**
Adds documentation to configuration of cache for webpack v5

Fixes #2605

### CODE REVIEW

File: src/content/configuration/other-options.md [hunk 1]
```diff
@@ -60,29 +60,167 @@ This will force webpack to exit its bundling process.
 
 `boolean` `object`
 
-Cache the generated webpack modules and chunks to improve build speed. Caching is enabled by default while in watch mode. To disable caching simply pass:
+Cache the generated webpack modules and chunks to improve build speed. `cache` is set to `type: 'memory'` in [`development` mode](/concepts/mode/#mode-development) and disabled in [`production` mode](/concepts/mode/#mode-production). To disable caching simply pass `false`:
```
Comment (sokra) [inline src/content/configuration/other-options.md]: Add a note that `cache: true` is an alias to `cache: { type: "memory" }`

Comment (sokra) [reply src/content/configuration/other-options.md]: Remove `simply` that makes the reader feel stupid.

File: src/content/configuration/other-options.md [hunk 2]
```diff
@@ -60,29 +60,167 @@ This will force webpack to exit its bundling process.
...
+  }
 };
 ```
 
-W> Don't share the cache between calls with different options.
+### `cache.hashAlgorithm`
+
+`string`
+
+Algorithm used the hash generation. See [Node.js crypto](https://nodejs.org/api/crypto.html) for more details. Defaults to `ml4`.
```
Comment (sokra) [inline src/content/configuration/other-options.md]: ml4 -> md4

File: src/content/configuration/other-options.md [hunk 3]
```diff
@@ -60,29 +60,167 @@ This will force webpack to exit its bundling process.
...
+```
+
+### `cache.loglevel`
+
+`string: 'debug' | 'info' | 'warning'`
+
+`cache.loglevel` tells webpack how much of `cache` log info to display.
+
+- `'debug'`: all access and errors with stack trace
+- `'info'`: all access
```
Comment (sokra) [inline src/content/configuration/other-options.md]: `verbose` was added in a later PR.

`'verbose'`: all access
`'info'`: all write access

File: src/content/configuration/other-options.md [hunk 4]
```diff
@@ -60,29 +60,167 @@ This will force webpack to exit its bundling process.
...
+- `'warning'`: only failed serialization
```
Comment (sokra) [inline src/content/configuration/other-options.md]: only failed serialization and deserialization

File: src/content/configuration/other-options.md [hunk 5]
```diff
@@ -60,29 +60,167 @@ This will force webpack to exit its bundling process.
...
+
+```javascript
+module.exports = {
+  //...
+  cache: {
+    type: 'filesystem',
+    loglevel: 'debug'
+  }
+};
+```
```
Comment (sokra) [inline src/content/configuration/other-options.md]: Note that the caching usually don't emit errors. All failures are warnings and fall back to not caching this item. The build will keep working. It makes sense to enable warnings if you want to investigate why caching isn't working / doesn't increase performance. It also makes sense when developing webpack plugins that affect caching.

When reporting bugs for the filesystem cache, make sure to get the stack trace of warnings with the `loglevel: "debug"` option.

File: src/content/configuration/other-options.md [hunk 6]
```diff
@@ -60,29 +60,167 @@ This will force webpack to exit its bundling process.
...
+
+```javascript
+module.exports = {
+  //...
+  cache: {
+    type: 'filesystem',
+    name: 'AppBuildCache'
+  }
+};
+```
```
Comment (sokra) [inline src/content/configuration/other-options.md]: Using `name` makes sense when you have multiple configurations which should have independent caches.

File: src/content/configuration/other-options.md [hunk 7]
```diff
@@ -60,29 +60,167 @@ This will force webpack to exit its bundling process.
...
+
+### `cache.store`
+
+`string: 'idle' | 'background' | 'instant'`
```
Comment (sokra) [inline src/content/configuration/other-options.md]: `'pack'` was added in a later PR.

File: src/content/configuration/other-options.md [hunk 8]
```diff
@@ -60,29 +60,167 @@ This will force webpack to exit its bundling process.
...
+
+`cache.store` tells webpack when to store data on the file system. Defaults to `'idle'`.
+
+- `'idle'`: Store data when compiler is idle
```
Comment (sokra) [inline src/content/configuration/other-options.md]: - `'pack'`: Store data when compiler is idle in a single file for all cached items
- `'idle'`: Store data when compiler is idle in one file per cached item

File: src/content/configuration/other-options.md [hunk 9]
```diff
@@ -60,29 +60,167 @@ This will force webpack to exit its bundling process.
...
 
 module.exports = {
   //...
-  cache: SharedCache
+  cache: {
+    type: 'filesystem',
+    cacheDirectory: path.resolve(__dirname, '.temp_cache')
+  }
 };
 ```
```
Comment (sokra) [inline src/content/configuration/other-options.md]: The final location of the cache is a combination of `cacheDirectory` + `name`.

File: src/content/configuration/other-options.md [hunk 10]
```diff
@@ -60,29 +60,167 @@ This will force webpack to exit its bundling process.
...
+    loglevel: 'debug'
+  }
+};
+```
+
+### `cache.name`
+
+`string: 'default'`
+
+Name for the cache. Different names will lead to different coexisting caches. Defaults to `'default'`.
```
Comment (sokra) [inline src/content/configuration/other-options.md]: The default is `` `${config.name}-${config.mode}` ``

config.name is the `name` option in the root of the config object.

File: src/content/configuration/other-options.md [hunk 11]
```diff
@@ -60,29 +61,174 @@ This will force webpack to exit its bundling process.
 
 `boolean` `object`
 
-Cache the generated webpack modules and chunks to improve build speed. Caching is enabled by default while in watch mode. To disable caching simply pass:
+Cache the generated webpack modules and chunks to improve build speed. `cache` is set to `type: 'memory'` in [`development` mode](/concepts/mode/#mode-development) and disabled in [`production` mode](/concepts/mode/#mode-production). `cache: true`is an alias to `cache: { type: 'memory' }`. To disable caching pass `false`:
```
Comment (sokra) [inline src/content/configuration/other-options.md]:
```suggestion
Cache the generated webpack modules and chunks to improve build speed. `cache` is set to `type: 'memory'` in [`development` mode](/concepts/mode/#mode-development) and disabled in [`production` mode](/concepts/mode/#mode-production). `cache: true` is an alias to `cache: { type: 'memory' }`. To disable caching pass `false`:
```

File: src/content/configuration/other-options.md [hunk 12]
```diff
@@ -60,29 +61,174 @@ This will force webpack to exit its bundling process.
...
+};
+```
+
+### `cache.store`
+
+`string: 'background' | 'idle' | 'instant' | 'pack'`
+
+`cache.store` tells webpack when to store data on the file system. Defaults to `'idle'`.
+
+- `'background'`: Store data in background while compiling, but doen't block the compilation
```
Comment (sokra) [inline src/content/configuration/other-options.md]:
```suggestion
- `'background'`: Store data in background while compiling, but doesn't block the compilation
```

File: src/content/configuration/other-options.md [hunk 13]
```diff
@@ -60,29 +61,174 @@ This will force webpack to exit its bundling process.
...
+`cache.store` option is only available when [`cache.type`](#cache-type) is set to `filesystem`.
+
+__webpack.config.js__
+
+```javascript
+module.exports = {
+  //...
+  cache: {
+    type: 'filesystem',
+    store: 'background'
```
Comment (sokra) [inline src/content/configuration/other-options.md]:
```suggestion
    store: 'pack'
```

background and instant make rarely sense, but pack is useful and boosts performance in many cases.

**Comment by EugeneHlushko** [comment]
Haha this suggest changes feature is amazing, but it would be nice to get it bulk

File: src/content/configuration/other-options.md (hunk 9 above)
Comment (EugeneHlushko) [reply src/content/configuration/other-options.md]: I've added a warning block two lines blow

File: src/content/configuration/other-options.md (hunk 8 above)
Comment (EugeneHlushko) [reply src/content/configuration/other-options.md]: changed

File: src/content/configuration/other-options.md (hunk 10 above)
Comment (EugeneHlushko) [reply src/content/configuration/other-options.md]: changed

File: src/content/configuration/other-options.md (hunk 7 above)
Comment (EugeneHlushko) [reply src/content/configuration/other-options.md]: added

File: src/content/configuration/other-options.md (hunk 4 above)
Comment (EugeneHlushko) [reply src/content/configuration/other-options.md]: updated

File: src/content/configuration/other-options.md (hunk 2 above)
Comment (EugeneHlushko) [reply src/content/configuration/other-options.md]: done

File: src/content/configuration/other-options.md (hunk 1 above)
Comment (EugeneHlushko) [reply src/content/configuration/other-options.md]: added a note and removed "simply" word

**Comment by EugeneHlushko** [comment]
Thanks!

**Pull request merged**